### PR TITLE
on post subject, check for duplicates from cache first

### PR DESCRIPTION
**Pull request by annyhe**

### CODE REVIEW

File: config/toggles.js [hunk 1]
```diff
@@ -126,6 +126,8 @@ const shortTermToggles = {
 
   returnUser: environmentVariableTrue(pe, 'RETURN_CREATEDBY_ON_TOKEN_INPUT'),
 
+  checkCacheOnPost: environmentVariableTrue(pe, 'CHECK_CACHE_ON_POST'),
+
```
Comment (shriramshankar) [inline config/toggles.js]: @annyhe @iamigo Do we have to do this behind a flag?

Comment (iamigo) [inline config/toggles.js]: Short-term, yes, just so we can control when we enable it in prod, and can turn it off if there's anything amiss.

File: api/v1/apiErrors.js [hunk 2]
```diff
@@ -42,6 +42,15 @@ apiErrors.create({
 });
 
 apiErrors.create({
+  code: 11102,
```
Comment (iamigo) [inline api/v1/apiErrors.js]: you can't re-use error code numbers

File: api/v1/controllers/subjects.js [hunk 3]
```diff
@@ -405,7 +405,34 @@ module.exports = {
    */
   postSubject(req, res, next) {
     validateRequest(req);
-    doPost(req, res, next, helper);
+    const { name, parentId, parentAbsolutePath } =
+      req.swagger.params.queryBody.value;
+
+    // if cache on and parentId is NOT provided
```
Comment (iamigo) [inline api/v1/controllers/subjects.js]: add a comment to explain why we're performing this check here

File: api/v1/controllers/subjects.js [hunk 4]
```diff
@@ -405,7 +405,34 @@ module.exports = {
...
+    if (featureToggles.isFeatureEnabled(sampleStoreConstants.featureName) &&
+      featureToggles.isFeatureEnabled('checkCacheOnPost') &&
+      !u.looksLikeId(parentId)) {
+
```
Comment (iamigo) [inline api/v1/controllers/subjects.js]: no empty line here

File: api/v1/controllers/subjects.js [hunk 5]
```diff
@@ -405,7 +405,34 @@ module.exports = {
...
+      const absolutePath = parentAbsolutePath ?
+        (parentAbsolutePath + '.' + name) : name;
+
```
Comment (iamigo) [inline api/v1/controllers/subjects.js]: no empty line here

File: api/v1/controllers/subjects.js [hunk 6]
```diff
@@ -405,7 +405,34 @@ module.exports = {
...
+      redisSubjectModel.isSubjectInCache(absolutePath)
+      .then((found) => {
+        if (found) {
+          throw new apiErrors.DuplicateResourceError(
+            'The subject lower case absolutePath must be unique');
+        } else {
```
Comment (iamigo) [inline api/v1/controllers/subjects.js]: no else required since you throw an error in the if

File: api/v1/controllers/subjects.js [hunk 7]
```diff
@@ -405,7 +405,34 @@ module.exports = {
...
+          doPost(req, res, next, helper);
+        }
+      })
+      .catch((err) => {
+        u.handleError(next, err, helper.modelName);
+      });
+    } else {
+
```
Comment (iamigo) [inline api/v1/controllers/subjects.js]: no empty line here

File: cache/models/subject.js [hunk 8]
```diff
@@ -39,6 +39,20 @@ const filters = {
 };
 
 /**
+ * Given absolutePath, return whether the subject is in cache
+ *
+ * @param {String} absolutePath
+ * @returns {Promise} resolves to true for found, false for not
+ */
+function isSubjectInCache(absolutePath) {
```
Comment (iamigo) [inline cache/models/subject.js]: boolean functions like this should take the form of a statement, not a question, so this function should be called "subjectIsInCache" or "subjectAlreadyExists" or "subjectExists" or "subjectInSampleStore"

File: cache/models/subject.js [hunk 9]
```diff
@@ -39,6 +39,20 @@ const filters = {
...
+  const subjectKey = sampleStore.toKey('subject', absolutePath);
+
+  // get from cache
+  return redisClient.sismemberAsync(sampleStore.constants.indexKey.subject, subjectKey)
+  .then((exist) => exist);
```
Comment (iamigo) [inline cache/models/subject.js]: Line 52 not needed. Just end line 51 with a semi-colon: you're returning a promise that resolves to 1 (truthy) if the subjectKey is already in the cache.

File: config/toggles.js (hunk 1 above)
Comment (iamigo) [inline config/toggles.js]: Please use "FAST_FAIL_DUPLICATE_SUBJECT" as the environment variable name, and "fastFailDuplicateSubject" as the toggle name.

File: tests/cache/models/subjects/post.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * tests/cache/models/subjects/post.js
+ */
+'use strict'; // eslint-disable-line strict
+
```
Comment (iamigo) [inline tests/cache/models/subjects/post.js]: no empty line here

File: tests/cache/models/subjects/post.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  before((done) => {
+    tu.toggleOverride('checkCacheOnPost', true);
+    tu.toggleOverride('enableRedisSampleStore', true);
+    tu.toggleOverride('getSubjectFromCache', true);
+    tu.createToken()
+    .then((returnedToken) => {
+      token = returnedToken;
+      done();
+    })
+    .catch((err) => done(err));
```
Comment (iamigo) [inline tests/cache/models/subjects/post.js]: .catch(done);

File: tests/cache/models/subjects/post.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+  });
+
+  beforeEach((done) => {
+    Subject.create(par)
+    .then(() => Subject.create({
+        name: childName,
```
Comment (iamigo) [inline tests/cache/models/subjects/post.js]: too many spaces

File: tests/cache/models/subjects/post.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  after(() => {tu.toggleOverride('checkCacheOnPost', false)});
+
+  it('no parent: duplicate name should fail from the cache', (done) => {
+    api.post(path)
+    .set('Authorization', token)
+    .send({ name: parentName })
+    .expect(constants.httpStatus.BAD_REQUEST)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/cache/models/subjects/post.js]: return done(err);

File: tests/cache/models/subjects/post.js [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+      }
+
+      expect(res.body.errors[0].type)
```
Comment (iamigo) [inline tests/cache/models/subjects/post.js]: if it fits on one line, keep on one line

File: tests/cache/models/subjects/post.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+  it('with parent: duplicate absolutePath should fail from the cache',
+    (done) => {
+    api.post(path)
+    .set('Authorization', token)
+    .send({ name: childName, parentAbsolutePath: parentName })
+    .expect(constants.httpStatus.BAD_REQUEST)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/cache/models/subjects/post.js]: return done(err);

File: tests/cache/models/subjects/post.js [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+      }
+
+      expect(res.body.errors[0].type)
+        .to.equal('DuplicateResourceError');
+      expect(res.body.errors[0].message)
+        .to.equal('The subject lower case absolutePath must be unique');
+      expect(res.body.errors[0].source)
+        .to.equal('Subject');
+
```
Comment (iamigo) [inline tests/cache/models/subjects/post.js]: no empty line here

File: tests/cache/models/subjects/post.js [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * tests/cache/models/subjects/post.js
+ */
```
Comment (shriramshankar) [inline tests/cache/models/subjects/post.js]: I know we have lots of API tests under tests/cache/models/subjects but wondering if we want to move the api tests to its own folder, under tests/cache/api/subject/... going forward. ?

File: tests/cache/models/subjects/post.js [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+const path = '/v1/subjects';
+const constants = require('../../../../api/v1/constants');
+const tu = require('../../../testUtils');
+const rtu = require('../redisTestUtil');
+const samstoinit = rtu.samstoinit;
+const redisClient = rtu.redisClient;
+const Subject = tu.db.Subject;
+const expect = require('chai').expect;
+
+describe('redis: subject: POST: ', () => {
```
Comment (shriramshankar) [inline tests/cache/models/subjects/post.js]: @iamigo @annyhe When tests fail, it is kind of difficult to find out which file the test is in. Any thoughts on adding the file location in the description instead of this. ?

File: tests/cache/models/subjects/post.js (hunk 17 above)
Comment (annyhe) [inline tests/cache/models/subjects/post.js]: we can do so in another PR :)

File: tests/cache/models/subjects/post.js (hunk 18 above)
Comment (annyhe) [inline tests/cache/models/subjects/post.js]: ok will do

Comment (iamigo) [inline tests/cache/models/subjects/post.js]: Yes, this was going to be next weekend's project -- the outermost "describe" for every test should be named for the file it's in (and every "describe" name should end with " >")
